text/0098-metric-instruments-explained.md

```diff
@@ -0,0 +1,207 @@
+
+[OTEP 93](https://github.com/open-telemetry/oteps/pull/93) proposed with a list of six standard instruments, the most necessary and useful combination of instrument refinements, plus one special case used to record timing measurements.  OTEP 93 was closed without merging after a more consistent approach to naming was uncovered.  [OTEP 96](https://github.com/open-telemetry/oteps/pull/96) made another proposal, that was closed in favor of this one after more debate surfaced.
+
+This proposal finalizes the naming proposal for the standard instruments, seeking to address core confusion related to the "Measure" and "Observer" terms:
+
+1. OTEP 88 stipulates that the terms currently in use to name synchronous and asynchronous instruments--"Measure" and "Observer"--become _abstract_ terms.  It also used phrases like "Measure-like" and "Observer-like" to discuss instruments with refinements.  This proposal states that we shall prefer the adjectives, commonly abbreviated "Sync" and "Async", when describing the kind of an instrument.  "Measure-like" means an instrument is synchronous.  "Observer-like" means that an instrument is asynchronous.
+2. There is inconsistency in the hypothetical naming scheme for instruments presented in OTEP 88.  Note that "Counter" and "Observer" end in "-er", a noun suffix used in the sense of "[person occupationally connected with](https://www.merriam-webster.com/dictionary/-er)", while the term "Measure" does not fit this pattern.  This proposal proposes to replace the abstract term "Measure" by "Recorder", since the associated function name (verb) is specified as `Record()`.
+
+This proposal also repeats the current specification--and the justification--for the default aggregation of each standard instrument.
+
+## Explanation
+
+The following table summarizes the final proposed standard instruments resulting from this set of proposals.  The columns are described in more detail below.
+
+| Existing name | **Standard name** | Instrument kind | Function name | Default aggregation | Measurement kind | Rate support (Monotonic) | Notes |
+| ------------- | ----------------------- | ----- | --------- | -------------- | ------------- | --- | ------------------------------------ |
+| Counter       | **Counter**             | Sync  | Add()     | Sum            | Delta         | Yes | Per-request, part of a monotonic sum |
+|               | **UpDownCounter**       | Sync  | Add()     | Sum            | Delta         | No  | Per-request, part of a non-monotonic sum |
+| Measure       | **ValueRecorder**       | Sync  | Record()  | MinMaxSumCount | Instantaneous | No  | Per-request, any non-additive measurement |
```

Comment:
I think this is Delta, because you want the min and max from in the interval (last export, now]. Also sum an more importantly count (which is a Counter) needs time association.

Reply:
(I'm composing an answer to this. I still believe this is a Instantaneous case, and will work to justify that reasoning.)

Reply:
Not convinced because you need to inform when you did a reset of the values. Would you expect to reset min/max/sum/count every export cycle?

Reply:
Yes, I would expect these values to "reset" every cycle, as they are instantaneous. I've written a more thorough explanation of what it means to aggregate `ValueObserver` results across both time and space dimensions.

Specifically, when there is no aggregation being applied, each ValueObserver outputs one value which is equal to the min, the max, and the sum, with a count of 1. 

See how you like this explanation: https://github.com/open-telemetry/oteps/pull/98/commits/299c442a9658a23333032d37cbd562ffb709019f

Reply:
Let's make sure we talk about the same thing. I am referring to **ValueRecorder** not ** ValueObserver**. Are we talking about the same thing?

Reply:
Also on the `ValueObserver` I do agree instantaneous is better.

Reply:
Oh, sorry. I misread your comment. We were talking about different instruments.

OK, so as for `ValueRecorder`, I see what you mean. I still think of the **values** as instantaneous measurements. I see how after aggregating measurements over an interval, the result of aggregation may take on a delta temporal quality, but I don't view that as automatic.

May we just say that the whenever aggregating results, an implementation must keep track of the time interval covered by the aggregation. After aggregating `ValueRecorder` measurements for several intervals, we'll have a distribution of instantaneous values covering a range of time, from which we can draw averages.

Reply:
I do agree that individual values are "instantaneous measurements" but the aggregated result is a "Delta" based on my understanding. This was one of the point from @jkwatson raised, that he needs the aggregation to reset every export (delta), and other backends prefer (cumulative). So we need to know in the exporter if this is delta or cumulative so I can do the right thing and build what I need for that exporter.

Reply:
Also in my mind every "measurement" that we record using sync instruments is "instantaneous" but the "Temporal" property apply to the aggregation (a.k.a the aggregated values were recorded in this interval).

Reply:
I've added to try to clarify my understanding here: https://github.com/open-telemetry/oteps/pull/98/commits/d2a90fe32df36d5d90112580f0799e634a3b9588

Reply:
I believe the thing you and @jkwatson are referring to will be a topic for the SDK specification.

Reply:
I don't think so, I would like to understand better then what the "temporal" property means in the proto. May be different concepts.

Reply:
Yes, I think it's important to distinguish between what kind of values the instruments *accept* vs. how they will be aggregated, and eventually exported. These are 3 different things.

Reply:
The question is: Is this "Delta" refer to the same "Delta" in the proto?

Reply:
My reading is 'no'. The "delta" here is a characteristic of the data being sent to the instrument. How it gets aggregated, and ends up the proto is a different thing.

Reply:
I just re-read https://github.com/open-telemetry/opentelemetry-proto/pull/140 and I don't see any disagreement here. The examples for Delta and Cumulative in the proto appear to agree with the usage here, in that it talks about the values being passed to the instrument.

I think I see where the confusion lies, but I don't think we want to change anything here. There are two ways in the proto that we can talk about something being cumulative. One: to say that a metric descriptor has the cumulative temporal quality, says something about the values it took in. Two: when we export a series of aggregations for a metric, continuing to use the same start time throughout the series. 

The first usage of cumulative refers to the kind of instrument: SumObserver and UpDownSumObserver are cumulative in this sense.

The second usage of cumulative is not the same: it refers to exported aggregations. Any exporter can be generate cumulative data for any kind of instrument, if it wants to (and has the memory to do so).

Example: I have a `ValueRecorder` and I'm measuring something non-additive (e.g., pedal position); these measurements are instantaneous.  A delta-oriented exporter for this instrument will output the distribution of instantaneous values for one interval (e.g., MMSC, DDSketch, T-digest, ...). A cumulative-oriented exporter for this instrument will output the distribution of instantaneous values since the beginning of time--still having the instantaneous temporal quality.

This will be taken up in the SDK specification: if an exporter wishes to convert a cumulative instrument to a cumulative aggregation, or a delta instrument to a delta aggregation, these are easy. To convert delta-to-cumulative or cumulative-to-delta requires more memory, and we can provide these configurations.

Reply:
Agreed. I was just saying that the value in the column here is not the value that will end up in the proto...that's something that depends on aggregation, view, or something else in the SDK.

The enumeration of possible temporal qualities may indeed end up the same, though. Just the mapping is not set by this OTEP, or by the API in general.

Reply:
Then this is confusing, because even the column name suggest that they are the same.

Reply:
These are the same concepts, though, used in separate places to describe inputs and outputs. It's OK with me, but if you'd like to suggest a way out (either here or in the proto), please do.

Reply:
Just to make sure I understand the output of the ValueRecorder if we would make a proto every export cycle will be (MinMaxSumCount, Delta), is that a correct understanding? And for a SumObserver will be (Int64 (assuming int), Cumulative)? What will be the output for ValueObserver?

If that is the case I would suggest to make sure we document that the column in this table refers to an interpretation of the input data and not the export data.

Reply:
@jmacd my understanding from the open-telemetry/opentelemetry-proto#140 is that it refers to the aggregated data, the kind of instrument is no longer important at that point, maybe this is a source of misunderstanding for me.

Reply:
Could we rename the `Temporal quality` column to `Input Temporal quality`, or maybe `Argument Temporal quality`?

I was initially confused into thinking this `Temporal quality` was describing the output quality, not the input data.

Is the `Rate Support (Monotonic)` column also describing input? I'm guessing not (it describes output), but want to check.

Reply:
Sorry, don't mean to add more work here. We can just make this clear in the eventual spec PR. I think it's important we all are on the same page conceptually here, but maybe the language can be a bit loose.

Reply:
I think its important to call out that the Temporal Quality in this table is NOT same as the one in Proto. Or use a different name if possible to avoid confusion.

I was trying to model this in my mind with this example below:

I am using ValueRecorder instrument to capture ServerLatency.

The following recordings occurred between times T1 and T2
valueRecorderInstrument.Record(100ms)
valueRecorderInstrument.Record(200ms)
valueRecorderInstrument.Record(50ms)
valueRecorderInstrument.Record(100ms)
valueRecorderInstrument.Record(5ms)

and following occurs between T2 and T3
valueRecorderInstrument.Record(1ms)
valueRecorderInstrument.Record(2000ms)
valueRecorderInstrument.Record(50ms)

The above calls are providing inputs, and the values are "instantaneous" inputs. This matches the temporal quality "Instantaneous" specified for ValueRecorder in the table above.

Now what I export would be one of the below (depending on my SDK/Batcher/Exporter)
{
 [Min 5 Max 200 Sum 500 Count 5, T1 to T2, TemporalQuality: Delta]
 [Min 1 Max 2000 Sum 2051 Count 3, T2 to T3, TemporalQuality: Delta]
}
or
{
[Min 1 Max 2000 Sum 2551 Count 8, T1 to T3, TemporalQuality: Delta]
}
or
{
 [Min 1 Max 2000 Sum 2551 Count 8, T1 onwards, TemporalQuality: Cumulative]
}

The TemporalQuality in the proto and instrument are not same. This is confusing for me. Would love to see this clarified!

Reply:
I don't see why we would use the cumulative vs. delta distinction in the way you're describing, and I also don't read this into the current protocol. I see the usage of these two terms as the same, they're adjectives, and they apply to inputs or to outputs of the metrics processing pipeline.

I would pass through the Instantaneous/Delta/Cumulative temporal quality of the original instrument in each case. Using Prometheus and Statsd as archetypes for the two kinds of exporter:

For Prometheus at T2:
{
[Min: 5, Max: 200, Sum: 455, Count: 5, TimeRange: T1 to T2, TemporalQuality: Instantaneous]
}

For Prometheus at T3:
{
[Min: 1, Max: 2000, Sum: 2506, Count: 8, TimeRange: T1 to T3, TemporalQuality: Instantaneous]
}

For Statsd at T2:
{
[Min: 5, Max: 200, Sum: 455, Count: 5, TimeRange: T1 to T2, TemporalQuality: Instantaneous]
}

For Statsd at T3:
{
[Min: 1, Max: 2000, Sum: 2051, Count: 3, TimeRange: T2 to T3, TemporalQuality: Instantaneous]
}

Reply:
Just for emphasis, I'm using a repeating start time to indicate that the aggregate result spans multiple collection periods. The temporal quality of export should match the instrument.

Reply:
> For Prometheus at T3:
> {
> [Min: 1, Max: 2000, Sum: 2506, Count: 8, TimeRange: T1 to T3, TemporalQuality: Instantaneous]
> }

Per my understanding, the above should be Cumulative. Proto file says for Instantaneous, starttime is not expected to be present (https://github.com/open-telemetry/opentelemetry-proto/blob/master/opentelemetry/proto/metrics/v1/metrics.proto#L165)

> For Statsd at T3:
> {
> [Min: 1, Max: 2000, Sum: 2051, Count: 3, TimeRange: T2 to T3, TemporalQuality: Instantaneous]
> }

Per my understanding, the above should be Delta. (https://github.com/open-telemetry/opentelemetry-proto/blob/master/opentelemetry/proto/metrics/v1/metrics.proto#L191)

I fully understand this is more for a SDK/Exporter topic, not API. I am happy with your stated purpose of this Otep - define standard instruments and their default aggregation.

Reply:
I filed this issue: https://github.com/open-telemetry/opentelemetry-proto/issues/147